enhancements/machine-config/on-cluster-builds.md

```diff
@@ -0,0 +1,1009 @@
+
+3.  The BuildController detects that these details are provided and
+    begins a build.
+
+4.  The built image is pushed to the cluster administrators’ image
+    registry of choice.
+
+5.  NodeController updates the `desiredImage` annotation on each node within its
+    pool according to its rules such as maximum unavailable nodes, which nodes
+    haven't been updated yet, etc.
+
+6.  The MachineConfigDaemon drains the target node, applies the new OS
+    content to the node, and reboots the node upon completion.
+
+#### Variation and form factor considerations
+
+##### Single-Node OpenShift (SNO)
+
+Single-Node OpenShift (SNO) should not have any difficulty running the
+on-cluster build process as-is. However, there may be some efficiency
+improvements that can be made. For example, the final image could be written
+directly to the nodes’ filesystem before the handoff to rpm-ostree / bootc to
+avoid a registry roundtrip.
+
+##### Hypershift
+
+Hypershift on the other hand, would require distinct configurations for
+both the host cluster and the guest cluster. Additionally, some
+adaptations would be required to Hypershift in order to traverse the
+host \<-\> guest cluster barrier. (need to page Hypershift MCO process
+back in to have a more complete opinion around this).
+
+##### Multi-Arch Clusters
+
+Multiarch clusters have special considerations in terms of needing to
+orchestrate more than one build for each target architecture as well as
+manifestlisting the final image. This consideration also extends to the image
+registry the final image will be pushed to. As of OpenShift 4.13, the on-cluster
+image registry now supports manifestlisted images, as does ImageStreams.
+
+##### Disconnected Environments
+
+Clusters in disconnected environments will likely require special environment
+configurations such as a local package mirror, local image registry, etc.
+However, on-cluster builds should continue to work as-is in these environments.
+
+##### FIPS-Enabled Environments
+
+FIPS-enabled environments have special requirements around ensuring that any
+cryptographic libraries used are certified. This may not impose any particular
+operational or implementation requirements for on-cluster builds, but it does
+require cluster admins to ensure that any custom OS content installed uses
+FIPS-compliant cryptographic libraries.
+
+### API Extensions
+
+#### Nodes (corev1.Node)
+
+The following annotations will be added to each node object to indicate
+the current and desired image states:
+
+| **Annotation Name:**                             | **Description:**                                                                       |
+| ------------------------------------------------ | -------------------------------------------------------------------------------------- |
+| `machineconfiguration.openshift.io/currentImage` | The fully-qualified image pullspec for the currently-booted OS image for a given node. |
```

Comment:
Can we move away from annotations on the node to separate CRDs tracking Node?

Reply:
This could eventually move to a node-adjacent CRD in the future. I think the development of that node-adjacent CRD is beyond the scope of on-cluster builds right now.